### PR TITLE
Small performance optimizations

### PR DESCRIPTION
**Pull request by filipsobol**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Other (engine): Small performance optimizations.
Other (utils): Performance optimizations of `getCallbacksForEvent` and `spliceArray`.

---

Improvements (already on top of #17444):

| Test              | Render time |
|-------------------|-------------|
| `paragraphs`      | -8.75%      |
| `lists`           | -2.44%      |
| `tableHuge`       | -10.61%     |
| `formattingLongP` | -14.63%     |
| `inlineStyles`    | -5.51%      |
| `mixed`           | -14.35%     |
| `wiki`            | -10.20%     |

### CODE REVIEW

File: packages/ckeditor5-engine/src/view/matcher.ts [hunk 1]
```diff
@@ -461,15 +465,13 @@ function matchAttributes(
...
-	if ( isPlainObject( patterns ) ) {
-		if ( ( patterns as any ).style !== undefined ) {
-			// Documented at the end of matcher.js.
-			logWarning( 'matcher-pattern-deprecated-attributes-style-key', patterns as any );
-		}
-		if ( ( patterns as any ).class !== undefined ) {
-			// Documented at the end of matcher.js.
-			logWarning( 'matcher-pattern-deprecated-attributes-class-key', patterns as any );
-		}
+	if ( ( patterns as any ).style !== undefined ) {
```
Comment (filipsobol) [inline packages/ckeditor5-engine/src/view/matcher.ts:468]: I'm not 100% sure about removing this `if` statement.

If it's needed, when add the following and remove the `as any` castings:

```ts
	if ( typeof patterns === 'object' && !( patterns instanceof RegExp ) && !Array.isArray( patterns ) ) {
```

File: packages/ckeditor5-engine/src/model/nodelist.ts [hunk 2]
```diff
@@ -247,13 +247,13 @@ export default class NodeList implements Iterable<Node> {
  * occupy multiple items if its offset size is greater than one.
  */
 function makeOffsetsArray( nodes: Array<Node> ): Array<Node> {
-	const offsets: Array<Node> = [];
+	const offsets = new Array<Node>();
```
Comment (niegowski) [inline packages/ckeditor5-engine/src/model/nodelist.ts:250]: Does this change anything?

File: packages/ckeditor5-engine/src/model/nodelist.ts [hunk 3]
```diff
@@ -247,13 +247,13 @@ export default class NodeList implements Iterable<Node> {
...
+	let index = 0;
 
 	for ( const node of nodes ) {
-		const start = offsets.length;
-
-		offsets.length += node.offsetSize;
-		offsets.fill( node, start );
+		for ( let i = 0; i < node.offsetSize; i++ ) {
```
Comment (niegowski) [inline packages/ckeditor5-engine/src/model/nodelist.ts:254]: Is `index` equal to `i`?

Comment (niegowski) [reply packages/ckeditor5-engine/src/model/nodelist.ts:254]: Ok, it's not.

File: packages/ckeditor5-engine/src/view/matcher.ts [hunk 4]
```diff
@@ -389,27 +386,34 @@ function matchPatterns(
...
 		} );
 	}
 
-	if ( isPlainObject( patterns ) ) {
-		return Object.entries( patterns );
+	if ( typeof patterns !== 'object' || patterns instanceof RegExp ) {
+		return [ [ patterns as any, true ] ];
+	}
+
+	// Below we do what Object.entries() does, but faster
```
Comment (niegowski) [inline packages/ckeditor5-engine/src/view/matcher.ts:407]:
```suggestion
	// Below we do what Object.entries() does, but faster.
```

File: packages/ckeditor5-engine/src/view/matcher.ts [hunk 5]
```diff
@@ -389,27 +386,34 @@ function matchPatterns(
...
 function normalizePatterns( patterns: PropertyPatterns ): Array<[ true | string | RegExp, true | string | RegExp ]> {
 	if ( Array.isArray( patterns ) ) {
-		return patterns.map( ( pattern: any ) => {
-			if ( isPlainObject( pattern ) ) {
-				if ( pattern.key === undefined || pattern.value === undefined ) {
-					// Documented at the end of matcher.js.
-					logWarning( 'matcher-pattern-missing-key-or-value', pattern );
-				}
+		return patterns.map( pattern => {
+			if ( typeof pattern !== 'object' || pattern instanceof RegExp ) {
```
Comment (scofalik) [inline packages/ckeditor5-engine/src/view/matcher.ts:390]: Wouldn't it be safer to check if it is a string?

> // Assume the pattern is either String or RegExp.

File: packages/ckeditor5-engine/src/view/matcher.ts (hunk 4 above)
Comment (scofalik) [reply packages/ckeditor5-engine/src/view/matcher.ts:407]: Again, surprised that writing something by hand is faster.

File: packages/ckeditor5-engine/src/model/nodelist.ts (hunk 2 above)
Comment (filipsobol) [reply packages/ckeditor5-engine/src/model/nodelist.ts:250]: No, it's a leftover from when I tried to specify array length, but decided to remove it as the improvement was minimal. I will change this to `[]`.

File: packages/ckeditor5-engine/src/view/matcher.ts [hunk 6]
```diff
@@ -389,27 +386,34 @@ function matchPatterns(
...
-			// Assume the pattern is either String or RegExp.
-			return [ pattern, true ];
+			return [ pattern.key, pattern.value ];
 		} );
 	}
 
-	if ( isPlainObject( patterns ) ) {
-		return Object.entries( patterns );
+	if ( typeof patterns !== 'object' || patterns instanceof RegExp ) {
+		return [ [ patterns as any, true ] ];
```
[the remark this answers is not in the document]
Comment (filipsobol) [reply packages/ckeditor5-engine/src/view/matcher.ts:404]: It's not. Removed.

File: packages/ckeditor5-engine/src/view/matcher.ts (hunk 5 above)
Comment (filipsobol) [reply packages/ckeditor5-engine/src/view/matcher.ts:390]: No, this check allows for either Regexp or string, not only string.

File: packages/ckeditor5-engine/src/model/nodelist.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (filipsobol) [reply packages/ckeditor5-engine/src/model/nodelist.ts:254]: Surprise, surprise. In the `table-huge` test, the `makeOffsetsArray` had the self-time and total time of ~34ms and ~51ms, respectively. After the changes, it's just ~0.8ms for both.

File: packages/ckeditor5-engine/src/view/matcher.ts (hunk 5 above)
Comment (scofalik) [reply packages/ckeditor5-engine/src/view/matcher.ts:390]: That's what I meant, `typeof pattern == 'string' || pattern instanceof RegExp` :smile:.

File: packages/ckeditor5-engine/src/conversion/downcastdispatcher.ts [hunk 7]
```diff
@@ -336,8 +336,8 @@ export default class DowncastDispatcher extends /* #__PURE__ */ EmitterMixin() {
 		}
 
 		// Fire a separate insert event for each node and text fragment contained in the range.
-		for ( const data of Array.from( range.getWalker( { shallow: true } ) ).map( walkerValueToEventData ) ) {
-			this._testAndFire( 'insert', data, conversionApi );
+		for ( const data of range.getWalker( { shallow: true } ) ) {
```
[the remark this answers is not in the document]
Comment (filipsobol) [reply packages/ckeditor5-engine/src/conversion/downcastdispatcher.ts:339]: Below you can find the improvements this change provides. I will let you and @niegowski be the judges of whether it is safe and worth it.

| Test name                    	| Before (ms) 	| After (ms) 	|
|------------------------------	|------------:	|-----------:	|
| `table-huge`                 	|          51 	|         14 	|
| `paragraphs`                 	|           9 	|          3 	|
| `lists`                      	|           8 	|        2.5 	|
| `formatting-long-paragraphs` 	|          31 	|        2.2 	|
| `inline-styles`              	|          21 	|        3.4 	|
| `mixed`                      	|          24 	|        7.2 	|
| `wiki`                       	|          22 	|        7.1 	|

File: packages/ckeditor5-engine/src/view/matcher.ts (hunk 5 above)
Comment (filipsobol) [reply packages/ckeditor5-engine/src/view/matcher.ts]: Right, makes sense

**Pull request merged**